### PR TITLE
feat(eslint-config): Update constructor access modifier eslint rule to require explicit modifiers

### PR DESCRIPTION
**Pull request by Josmithr**
Updates rule to match discussion outcome in [this discussion](https://github.com/microsoft/FluidFramework/pull/9502#discussion_r828479192).

Rule documentation: https://typescript-eslint.io/rules/explicit-member-accessibility/

Also fixes violations in packages using the `strict` config

### Breaking Change

Note that this is a breaking change with respect to the above rule. We are changing error-level enforcement of the policy to the opposite rule, so code changes will be required. This PR pre-fixes the violations existing at this point in time, but any changes between now and when the new package is published and adopted will need to be fixed.

### CODE REVIEW

File: packages/drivers/local-driver/src/localSessionStorageDb.ts [hunk 1]
```diff
@@ -16,6 +16,12 @@ class LocalSessionStorageCollection<T> implements ICollection<T> {
      * @param collectionName - data type of the collection, e.g. blobs, deltas, trees, etc.
      */
     constructor(private readonly collectionName: string) { }
+    async findAndUpdate(query: any, value: T): Promise<{ value: T; existing: boolean; }> {
+        throw new Error("Method not implemented.");
+    }
+    createTTLIndex?(index: any, mongoExpireAfterSeconds?: number | undefined): Promise<void> {
+        throw new Error("Method not implemented.");
+    }
```
Comment (alexvy86) [inline packages/drivers/local-driver/src/localSessionStorageDb.ts:24]: Stowaways in this PR? 😄

**Review by alexvy86 — COMMENTED** [review]
Looks good to me other than the single comment below.

File: packages/drivers/local-driver/src/localSessionStorageDb.ts (hunk 1 above)
Comment (Josmithr) [reply packages/drivers/local-driver/src/localSessionStorageDb.ts:24]: Indeed so. 🤦‍♂️ Thanks!

**Review by ChumpChief — APPROVED** [review]
aw yeah

**Comment by Josmithr** [comment]
Waiting on ongoing discussion about major version changes in this and other individually published packages.

File: azure/packages/azure-client/.eslintrc.js [hunk 2]
```diff
@@ -31,6 +31,21 @@ module.exports = {
         // Ensure jsdoc/tsdoc comments contain a main description component
         // (disallows empty comments / only tags).
         "jsdoc/require-description": ["error", { checkConstructors: false }],
+
+        // TODO: remove once dependency on base config has been updated.
+        "@typescript-eslint/explicit-member-accessibility": [
```
Comment (CraigMacomber) [inline azure/packages/azure-client/.eslintrc.js:36]: It's unclear to me that checking this in is worth while. Doing so just to remove it in a few days seems low value.

Might be better to test with npm-link to find cases with violate the version of the lint config in the repo.

Comment (Josmithr) [reply azure/packages/azure-client/.eslintrc.js:36]: That's what these changes represent. They are pre-fixing to accommodate the new rule, but that new rule violates the existing rules, so unless we change the local configs to match (for now), then linting will fail.

File: packages/dds/quorum/.eslintrc.js [hunk 3]
```diff
@@ -5,10 +5,25 @@
 
 module.exports = {
     "extends": [
-        "@fluidframework/eslint-config-fluid/strict",
+        require.resolve("@fluidframework/eslint-config-fluid/strict"),
```
Comment (CraigMacomber) [inline packages/dds/quorum/.eslintrc.js:8]: this seems like an unrelated change. Also, why do we want to use require.resolve here, but not for "prettier"?



There are several config files that do it the way this one does (my regex suggests there are 17 such configs).
Maybe the simpler version without require.resolve works?

My understanding of what require.resolve does suggests that using it in a top level config file thats only used as part of the project its located in makes it pointless since it shouldn't always work fine without it.

**Review by ChumpChief — APPROVED** [review]

File: tools/api-markdown-documenter/prettier.config.cjs [hunk 4]
```diff
@@ -5,6 +5,7 @@
 
 module.exports = {
     ...require("@fluidframework/build-common/prettier.config.cjs"),
+    useTabs: false, // @fluidframework/build-common ^1.2.0, change to TRUE after prettier infra
```
Comment (CraigMacomber) [inline tools/api-markdown-documenter/prettier.config.cjs:8]: this seems unrelated.

Comment (CraigMacomber) [reply tools/api-markdown-documenter/prettier.config.cjs:8]: also comment looks truncated?

Comment (Josmithr) [reply tools/api-markdown-documenter/prettier.config.cjs:8]: I made a note about this in the PR description, but extracting this into its own PR is probably better. I'll do that.

File: packages/dds/quorum/.eslintrc.js (hunk 3 above)
Comment (Josmithr) [reply packages/dds/quorum/.eslintrc.js:8]: eslint's default module resolution results in transitive deps not beingrespected. I.e. for a config extension chain like `a -(extends)->b->c`, `a` is required to directly depend on `c`. This applies to all "extends" and "plugin" entries. Directly `require`ing `b` from `a` lets node resolve the dependencies, and ends up yielding the behavior you would generally expect from a node tool.

This pattern is a pretty widely used workaround. I'll have to see if I can dig up the github issue on this.

It's possible that hoisting the require calls up into our shared config might also work, and would remove the requirement on the consumer side. I would have to test it.

Either way, these changes could be pulled into a separate PR, so that's probably the best call.

File: azure/packages/azure-client/.eslintrc.js (hunk 2 above)
Comment (CraigMacomber) [reply azure/packages/azure-client/.eslintrc.js:36]: so the current default bans explicit constructor public scope instead of just not requiring it?

File: packages/dds/quorum/.eslintrc.js (hunk 3 above)
Comment (Josmithr) [reply packages/dds/quorum/.eslintrc.js:8]: This PR has more details (and is where these changes were initially made): #10410

File: azure/packages/azure-client/.eslintrc.js (hunk 2 above)
Comment (Josmithr) [reply azure/packages/azure-client/.eslintrc.js:36]: Correct.

File: packages/dds/quorum/.eslintrc.js (hunk 3 above)
Comment (CraigMacomber) [reply packages/dds/quorum/.eslintrc.js:8]: My understanding is the shared config should do `require.resolve` because it might refer to packages which don't get hoisted into the root, and thus won't be found without it, unless you get lucky and npm happened to hoist those deps.

In general, code which is distributed with a package, and needs to generate string paths to packages which it passes off to some other library should use require.resolve instead of unqualified strings.

Doing it in this file, which is not distributed as part of its package seems pointless. This file is always used when running eslint inside this packages' directory, and since this package declares a dependency on the referenced package, the package manager (npm for now) will ensure this path is in scope for the only case that its used.

Comment (CraigMacomber) [reply packages/dds/quorum/.eslintrc.js:8]: https://github.com/eslint/eslint/issues/3458#issuecomment-132922673 notes that it's the shred config that should use module objects. This justification does not apply to non-shared configs like this one. I think this approach was misapplied in your previous change you linked.

Comment (Josmithr) [reply packages/dds/quorum/.eslintrc.js]: Yeah, I think that is the better approach. At this point I don't remember why it wasn't the approach I ended up taking. I'll have to do some testing.

Note that this pattern did allow us to remove the dependencies on transitive configs, so it does end up having an impact in at least some cases, but I'll have to see if I can still repro. 

I have removed this change from the PR, though, so I'm going to resolve this issue for now. I'll keep exploring though :)

File: tools/api-markdown-documenter/prettier.config.cjs (hunk 4 above)
Comment (Josmithr) [reply tools/api-markdown-documenter/prettier.config.cjs]: 🤦‍♂️This was only required because I was in symlinked mode... Changes reverted.

File: packages/framework/fluid-framework/.eslintrc.js [hunk 5]
```diff
@@ -4,9 +4,7 @@
  */
 
 module.exports = {
-    "extends": [
-        "@fluidframework/eslint-config-fluid/strict", "prettier"
-    ],
+    "extends": ["@fluidframework/eslint-config-fluid/strict", "prettier"],
```
Comment (CraigMacomber) [inline packages/framework/fluid-framework/.eslintrc.js:7]: unrelated formatting change here

**Review by CraigMacomber — APPROVED** [review]

**Review by CraigMacomber — APPROVED** [review]
